unidist/core/backends/mpi/core/shared_object_store.py

```diff
@@ -0,0 +1,775 @@
+# Copyright (C) 2021-2023 Modin authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""`SharedObjectStore` functionality."""
+
+import os
+import sys
+import time
+import psutil
+import weakref
+from array import array
+
+from cmemory import write_to
+
+try:
+    import mpi4py
+except ImportError:
+    raise ImportError(
+        "Missing dependency 'mpi4py'. Use pip or conda to install it."
+    ) from None
+
+from unidist.config.backends.mpi.envvars import (
+    MpiSharedObjectStoreMemory,
+    MpiSharedObjectStoreThreshold,
+    MpiBackoff,
+)
+from unidist.core.backends.mpi.core import common, communication
+from unidist.core.backends.mpi.core.serialization import ComplexDataSerializer
+
+# TODO: Find a way to move this after all imports
+mpi4py.rc(recv_mprobe=False, initialize=False)
+from mpi4py import MPI  # noqa: E402
+
+
+class SharedSignaler:
+    """
+    Class that help to synchronize write to shared memory.
+
+    Parameters
+    ----------
+    win : MPI.win
+        The MPI window that was used to create the shared memory.
+
+    Notes
+    -----
+    This class should be used with the `with` statement.
+    """
+
+    def __init__(self, win):
+        self.win = win
+
+    def __enter__(self):
+        """Lock the current MPI.Window for other processes."""
+        self.win.Lock(communication.MPIRank.MONITOR)
+
+    def __exit__(self, *args):
+        """Unlock the current MPI.Window for other processes."""
+        self.win.Unlock(communication.MPIRank.MONITOR)
+
+
+class SharedObjectStore:
+    """
+    Class that provides access to data in shared memory
+
+    Notes
+    -----
+    This class initializes and manage shared memory.
+    """
+
+    __instance = None
+
+    # Service constants defining the structure of the service buffer
+    SERVICE_COUNT = 20000
```

Comment:
Let's add a config value for this parameter like MpiServiceCount. Rename later during the review.